### PR TITLE
multi: implement latest variable sized EOB proposal

### PR DESCRIPTION
**Pull request by Roasbeef**
This picks up where #31, #33, and #36 left off. It implements the latest version of  lightningnetwork/lightning-rfc#619. The only thing wrong with this PR atm is that it fails the EOB JSON test included in the latest spec PR, while passing end-to-end tests within the same implementation.

### CODE REVIEW

**Comment by Roasbeef** [comment]
Fixed some issues mapping the JSON test vectors, we now match up with the spec's test vectors!

**Comment by Roasbeef** [comment]
PTAL @cfromknecht 

(won't build until the TLV PR in lnd is merged)

**Comment by Roasbeef** [comment]
Rebased now to the latest version of `lnd` that has the `tlv` package merged.

File: sphinx.go [hunk 1]
```diff
@@ -44,6 +44,25 @@ const (
 	HopDataSize = (RealmByteSize + AddressSize + AmtForwardSize +
 		OutgoingCLTVSize + NumPaddingBytes + HMACSize)
 
+	// FramSize is the size of a frame in the packet. A frame is a region
```
Comment (cfromknecht) [inline sphinx.go]: s/FramSize/FrameSize

File: sphinx.go [hunk 2]
```diff
@@ -52,13 +71,14 @@ const (
...
 	// used, then the remainder is padded with null-bytes, also obfuscated.
-	routingInfoSize = NumMaxHops * HopDataSize
+	routingInfoSize = NumMaxHops * FrameSize
 
 	// numStreamBytes is the number of bytes produced by our CSPRG for the
 	// key stream implementing our stream cipher to encrypt/decrypt the mix
-	// header. The last hopDataSize bytes are only used in order to
-	// generate/check the MAC over the header.
-	numStreamBytes = routingInfoSize + HopDataSize
+	// header. The maxPayloadSize bytes at the end are used to
```
Comment (cfromknecht) [inline sphinx.go]: s/maxPayloadSize/MaxPayloadSize

File: path.go [hunk 3]
```diff
@@ -1,9 +1,313 @@
...
+	err = binary.Read(r, binary.BigEndian, &hd.OutgoingCltv)
+	if err != nil {
+		return err
+	}
+
+	if _, err := io.ReadFull(r, hd.ExtraBytes[:]); err != nil {
+		return err
+	}
+
+	return err
```
Comment (cfromknecht) [inline path.go]: return nil, even tho we can only get here if `err` was nil above

File: path.go [hunk 4]
```diff
@@ -1,9 +1,313 @@
...
+		return err
+	}
+
+	err = binary.Read(r, binary.BigEndian, &hd.OutgoingCltv)
+	if err != nil {
+		return err
+	}
+
+	if _, err := io.ReadFull(r, hd.ExtraBytes[:]); err != nil {
+		return err
```
Comment (cfromknecht) [inline path.go]: could instead return `err` directly

File: path.go [hunk 5]
```diff
@@ -1,9 +1,313 @@
...
+		varInt, err := tlv.ReadVarInt(bufReader, &b)
+		if err != nil {
+			return err
+		}
+
+		payloadSize = uint32(varInt)
+		hp.Type = PayloadTLV
+	}
+
+	// Now that we now the payload size, we'll a new buffer to read it out
```
Comment (cfromknecht) [inline path.go]: Now that we know*
we'll create(?) a new buffer

File: path_test.go [hunk 6]
```diff
@@ -0,0 +1 @@
+package sphinx
```
Comment (cfromknecht) [inline path_test.go]: extraneous file

File: path.go [hunk 7]
```diff
@@ -1,9 +1,313 @@
...
+	// hop payload structure from the payload.
+	if hp.Type != PayloadLegacy {
+		return nil, nil
+	}
+
+	// Now that we know the payload has the structure we expect, we'll
+	// decode the payload into the HopData.
+	var hd HopData
+	if err := hd.Decode(payloadReader); err != nil {
+		return &hd, err
```
Comment (cfromknecht) [inline path.go]: is returning `&hd` intentional?

File: path.go [hunk 8]
```diff
@@ -1,9 +1,313 @@
...
+	// includes the higher-level (optional) associated data bytes.
+	HMAC [HMACSize]byte
+}
+
+// NewHopPayload creates a new hop payload given an optional set of forwarding
+// instructions for a hop, and a set of optional opaque extra onion bytes to
+// drop off at the target hop. If both values are not specified, then an error
+// is returned.
+func NewHopPayload(hopData *HopData, eob []byte) (HopPayload, error) {
+
```
Comment (cfromknecht) [inline path.go]: nit: extra new line

File: path.go [hunk 9]
```diff
@@ -1,9 +1,313 @@
...
+		h HopPayload
+		b bytes.Buffer
+	)
+
+	// We can't proceed if neither the hop data or the EOB has been
+	// specified by the caller.
+	if hopData == nil && len(eob) == 0 {
+		return h, fmt.Errorf("either hop data or eob must " +
+			"be specified")
+	}
```
Comment (cfromknecht) [inline path.go]: should also fail if `hopData != nil && len(eob) > 0`?

File: sphinx.go (hunk 1 above)
Comment (cfromknecht) [reply sphinx.go]: can ignore this one

File: path.go (hunk 9 above)
Comment (Roasbeef) [reply path.go]: Yep! I wan to force the caller to only provide one. In retrospect, I think I favor an API where you provide one of the other, but it was done this way in order to minimize code impact in the router within `lnd`. Once this goes in, I may follow up with a partial re-design of the interface.

File: path.go (hunk 7 above)
Comment (Roasbeef) [reply path.go]: Nope, fixed!

**Comment by Roasbeef** [comment]
PTAL @cfromknecht

**Pull request merged**